### PR TITLE
feat(compiler): Add validation for introspection fields in subscription operations

### LINKED ISSUE #305: validation: subscription operation must not have an introspection field in selection set
check that fields in subscription selection set are not introspection fields. [spec](https://spec.graphql.org/October2021/#sel-FALPHDGDDLAAADFBA04c)

### PR DESCRIPTION
**Pull request by gocamille**
### Description:
This PR adds validation for introspection fields in subscription operations in the apollo-compiler library. The code checks that the root field of a subscription selection is not an introspection field, and returns validation errors if the root field is an introspection field.

Fixes #305 .

### CODE REVIEW

File: crates/apollo-compiler/src/validation/subscription.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+                }
+            }
+        }
+    }
+
+    diagnostics
+}
+
+#[cfg(test)]
+
```
Comment (gocamille) [inline crates/apollo-compiler/src/validation/subscription.rs:33]: @lrlna I kept this module-level test in the module, but should it be removed for the sake of convention and consistency?

File: crates/apollo-compiler/test_data/diagnostics/0050_subscription_operation_with_root_introspection_field.graphql [hunk 2]
```diff
@@ -0,0 +1,3 @@
+subscription sub {
+  __typename
+}
```
Comment (lrlna) [inline crates/apollo-compiler/test_data/diagnostics/0050_subscription_operation_with_root_introspection_field.graphql:3]: Let's add a subscription type to this document so we don't also get a `UnsupportedOperation` diagnostic in our output

```suggestion
subscription sub {
  __typename
}

type Subscription {
  __typename: String
}
```

File: crates/apollo-compiler/src/validation/subscription.rs (hunk 1 above)
Comment (lrlna) [reply crates/apollo-compiler/src/validation/subscription.rs:33]: yes, let's keep it these sort of tests in `test_data` directory.

File: crates/apollo-compiler/src/validation/subscription.rs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    let mut diagnostics = Vec::new();
+    let subscription_operations = db.upcast().subscription_operations(file_id);
+
+    // Subscription fields must not have an introspection field in the selection set.
+    //
+    // Return an Introspection Field error in case of an introspection field existing as the root field in the set.
+    for op in subscription_operations.iter() {
+        for selection in op.selection_set().selection() {
+            if let Selection::Field(field) = selection {
+                if field.name().starts_with("__") {
```
Comment (lrlna) [inline crates/apollo-compiler/src/validation/subscription.rs:15]: it's only a convention to keep `__` for introspection fields only, so theoretically someone can just have a field that starts with `__` in their schema. It'd be best to check for `__typename`, `__schema` and `__type` directly. 

If you rebase from main, we've just merged #421, so you should be able to use `field.is_introspection()` here.

File: crates/apollo-compiler/src/diagnostics.rs [hunk 4]
```diff
@@ -425,3 +428,20 @@ pub struct UnsupportedLocation {
...
+
+#[derive(Diagnostic, Debug, Error, Clone, Hash, PartialEq, Eq)]
+#[error("Subscription operations can not have an introspection field as a root field")]
+#[diagnostic(code("apollo-compiler validation error"))]
+pub struct IntrospectionField {
+    // current definition
+    pub field: String,
+
+    #[help]
+    pub help: Option<String>,
```
Comment (lrlna) [inline crates/apollo-compiler/src/diagnostics.rs:440]: i think `help` field is not necessary here. The `error` above describes the problem, and the label for definition describes what's wrong with this field.

File: crates/apollo-compiler/src/diagnostics.rs [hunk 5]
```diff
@@ -425,3 +428,20 @@ pub struct UnsupportedLocation {
     #[help]
     pub help: Option<String>,
 }
+
+#[derive(Diagnostic, Debug, Error, Clone, Hash, PartialEq, Eq)]
+#[error("Subscription operations can not have an introspection field as a root field")]
```
Comment (lrlna) [inline crates/apollo-compiler/src/diagnostics.rs:433]: going to try to stick with Rust's error messaging conventions, and keep things lower cased 😅 

```suggestion
#[error("subscription operations can not have an introspection field as a root field")]
```

**Review by lrlna — COMMENTED** [review]
This is looking good!! I'd rebase from main, as we just merged #421, which will be helpful for this particular validation rule.

File: crates/apollo-compiler/src/validation/subscription.rs (hunk 3 above)
Comment (gocamille) [reply crates/apollo-compiler/src/validation/subscription.rs:15]: Ah, good to know! I'll update this and use that great `field.is_introspection()` method 👍

**Pull request merged**